From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kenified <kenified@protonmail.com>
Date: Mon, 9 Jan 2023 11:53:15 +0100
Subject: [PATCH] More falling blocks


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 2a22d0639d63a3db4e7bded2d4e91b09bb93a95e..38dc1d779533332dc3daec864e46673dd89514a7 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -8269,6 +8269,15 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case SAND:
             case WHITE_CONCRETE_POWDER:
             case YELLOW_CONCRETE_POWDER:
+            case GRASS_BLOCK:
+            case DIRT:
+            case COARSE_DIRT:
+            case PODZOL:
+            case FARMLAND:
+            case SNOW_BLOCK:
+            case CLAY:
+            case MYCELIUM:
+            case DIRT_PATH:
             // ----- Legacy Separator -----
             case LEGACY_SAND:
             case LEGACY_GRAVEL:
