From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kenified <kenified@protonmail.com>
Date: Mon, 9 Jan 2023 11:53:15 +0100
Subject: [PATCH] More falling blocks


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 4df3de56641b0ce028f25ee8d3ed496c707302a7..de71f2208c2bb6a70d66f55b2d219027f665ee9c 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -8277,6 +8277,15 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
             case SAND:
             case WHITE_CONCRETE_POWDER:
             case YELLOW_CONCRETE_POWDER:
+            case GRASS_BLOCK:
+            case DIRT:
+            case COARSE_DIRT:
+            case PODZOL:
+            case FARMLAND:
+            case SNOW_BLOCK:
+            case CLAY:
+            case MYCELIUM:
+            case DIRT_PATH:
             // ----- Legacy Separator -----
             case LEGACY_SAND:
             case LEGACY_GRAVEL:
